go/parquet/s3/s3_reader.go

```diff
@@ -0,0 +1,305 @@
+// Optionally the file can be memory mapped for faster reading. If no read properties are provided
+// then the default ReaderProperties will be used. The WithMetadata option can be used to provide
+// a FileMetaData object rather than reading the file metadata from the file.
+func OpenParquetFile(filename string, opts ...ReadOption) (*Reader, error) {
+	ctx := context.Background()
+	bucket := "writo-test-bucket"
+
+	// var source source.ParquetFile
+	source, err := s3.NewS3FileReader(ctx, bucket, filename)
+	if err != nil {
+		return nil, err
+	}
+	src := S3file { source }
+	return NewParquetReader(src, opts...)
+}
+
+// NewParquetReader returns a FileReader instance that reads a parquet file which can be read from r.
+// This reader needs to support Read, ReadAt and Seeking.
+//
+// If no read properties are provided then the default ReaderProperties will be used. The WithMetadata
+// option can be used to provide a FileMetaData object rather than reading the file metadata from the file.
+func NewParquetReader(r parquet.ReaderAtSeeker, opts ...ReadOption) (*Reader, error) {
+	var err error
+	f := &Reader{r: r}
+	for _, o := range opts {
+		o(f)
+	}
+
+	if f.footerOffset <= 0 {
+		f.footerOffset, err = r.Seek(0, io.SeekEnd)
+		if err != nil {
+			return nil, fmt.Errorf("parquet: could not retrieve footer offset: %w", err)
+		}
+	}
+
+	if f.props == nil {
+		f.props = parquet.NewReaderProperties(memory.NewGoAllocator())
+	}
+
+	f.bufferPool = sync.Pool{
+		New: func() interface{} {
+			buf := memory.NewResizableBuffer(f.props.Allocator())
+			runtime.SetFinalizer(buf, func(obj *memory.Buffer) {
+				obj.Release()
+			})
+			return buf
+		},
+	}
+
+	f.footerOffset = 0
+	if f.metadata == nil {
+		return f, f.parseMetaData()
+	}
+
+	return f, nil
+}
+
+// BufferPool returns the internal buffer pool being utilized by this reader.
+// This is primarily for use by the pqarrow.FileReader or anything that builds
+// on top of the Reader and constructs their own ColumnReaders (like the
+// RecordReader)
+func (f *Reader) BufferPool() *sync.Pool {
+	return &f.bufferPool
+}
+
+// Close will close the current reader, and if the underlying reader being used
+// is an `io.Closer` then Close will be called on it too.
+func (f *Reader) Close() error {
+	if r, ok := f.r.(io.Closer); ok {
+		return r.Close()
+	}
+	return nil
+}
+
+// MetaData returns the underlying FileMetadata object
+func (f *Reader) MetaData() *metadata.FileMetaData { return f.metadata }
+
+// parseMetaData handles parsing the metadata from the opened file.
+func (f *Reader) parseMetaData() error {
+	if f.footerOffset <= int64(footerSize) {
+		return fmt.Errorf("parquet: file too small (size=%d)", f.footerOffset)
+	}
+
+	buf := make([]byte, footerSize)
+	// backup 8 bytes to read the footer size (first four bytes) and the magic bytes (last 4 bytes)
+	n, err := f.r.ReadAt(buf, f.footerOffset-int64(footerSize))
+	if err != nil {
+		return fmt.Errorf("parquet: could not read footer: %w", err)
+	}
```

Comment:
All of this logic already exists in the `file` package of the Parquet module and is unnecessary to reproduce. All you need to do is have something that meets the `parquet.ReaderAtSeeker` interface and everything else will be handled automatically.